### PR TITLE
Added 'Reverse' option to StackTraceLayoutRenderer

### PR DESCRIPTION
**Pull request by AlanLiu90**

### CODE REVIEW

File: src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs [hunk 1]
```diff
@@ -146,65 +161,106 @@ private static void AppendRaw(StringBuilder builder, LogEventInfo logEvent, int
...
-                }
-                else
-                {
-                    builder.Append("<no type>");
-                }
-
-                builder.Append(".");
-                builder.Append(method.Name);
-                first = false;
+                AppendFlat(builder, logEvent, i, ref first);
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs:187]: Instead of having first as `ref`-parameter then maybe do like this:

```
if (AppendFlat(builder, logEvent, i, first))
    first = false;
```

File: src/NLog/SetupExtensionsBuilderExtensions.cs [hunk 2]
```diff
@@ -142,9 +144,18 @@ public static ISetupExtensionsBuilder RegisterLayoutRenderer(this ISetupExtensio
         /// <param name="setupBuilder">Fluent interface parameter.</param>
         /// <param name="name">Name of the layout renderer - without ${}.</param>
         /// <param name="layoutMethod">Callback that returns the value for the layout renderer.</param>
-        public static ISetupExtensionsBuilder RegisterLayoutRenderer(this ISetupExtensionsBuilder setupBuilder, string name, Func<LogEventInfo, LoggingConfiguration, object> layoutMethod)
+        /// <param name="options">Options of the layout renderer.</param>
+        public static ISetupExtensionsBuilder RegisterLayoutRenderer(this ISetupExtensionsBuilder setupBuilder, string name, Func<LogEventInfo, LoggingConfiguration, object> layoutMethod, LayoutRenderOptions options = LayoutRenderOptions.None)
```
Comment (snakefoot) [inline src/NLog/SetupExtensionsBuilderExtensions.cs:148]: Think this is a good change. But can it be moved to another PR? And by adding new methods with extra parameter. Since changing existing is breaking change.

File: src/NLog/SetupExtensionsBuilderExtensions.cs [hunk 3]
```diff
@@ -142,9 +144,18 @@ public static ISetupExtensionsBuilder RegisterLayoutRenderer(this ISetupExtensio
...
         {
-            var layoutRenderer = new FuncLayoutRenderer(name, layoutMethod);
+            FuncLayoutRenderer layoutRenderer;
+
+            if ((options & LayoutRenderOptions.ThreadAgnostic) == LayoutRenderOptions.ThreadAgnostic)
```
Comment (snakefoot) [inline src/NLog/SetupExtensionsBuilderExtensions.cs:152]: Possible something could be done to reuse the logic from Layout.CreateFuncLayoutRenderer. Not good with duplicate code.

File: src/NLog/SetupExtensionsBuilderExtensions.cs (hunk 2 above)
Comment (AlanLiu90) [reply src/NLog/SetupExtensionsBuilderExtensions.cs:148]: How can I move it to another PR? It seems that when I push my local commit to Github, it's added to this PR automatically.

Comment (snakefoot) [reply src/NLog/SetupExtensionsBuilderExtensions.cs:148]: Right now you are making a PullRequest from your own local `dev`-branch. If you perform `git branch` into another new local-branch, then it can be used for creating another seperate PullRequest from the new local-branch.

**Comment by snakefoot** [comment]
Instead of creating duplicate code with reverse-iteration, then maybe something like this:

```
        /// <summary>
        /// Gets or sets whether to render StackFrames in reverse order
        /// </summary>
        /// <docgen category='Rendering Options' order='10' />
        [DefaultValue(false)]
        public bool Reverse { get; set; }

        protected override void Append(StringBuilder builder, LogEventInfo logEvent)
        {
            if (logEvent.StackTrace == null)
                return;

            int startingFrame = logEvent.UserStackFrameNumber + TopFrames - 1;
            if (startingFrame >= logEvent.StackTrace.GetFrameCount())
            {
                startingFrame = logEvent.StackTrace.GetFrameCount() - 1;
            }

            int endingFrame = logEvent.UserStackFrameNumber + SkipFrames;
            StackFrameList stackFrameList = new StackFrameList(logEvent.StackTrace, startingFrame, endingFrame, Reverse);

            switch (Format)
            {
                case StackTraceFormat.Raw:
                    AppendRaw(builder, stackFrameList);
                    break;

                case StackTraceFormat.Flat:
                    AppendFlat(builder, stackFrameList);
                    break;

                case StackTraceFormat.DetailedFlat:
                    AppendDetailedFlat(builder, stackFrameList);
                    break;
            }
        }

        private struct StackFrameList : IEnumerable<StackFrame>
        {
            private readonly StackTrace _stackTrace;
            private readonly int _startingFrame;
            private readonly int _endingFrame;
            private readonly bool _reverse;

            public int Count => _startingFrame - _endingFrame;

            public StackFrame this[int index]
            {
                get
                {
                    int orderedIndex = _reverse ? _endingFrame + index + 1 : _startingFrame - index;
                    return _stackTrace.GetFrame(orderedIndex);
                }
            }

            public StackFrameList(StackTrace stackTrace, int startingFrame, int endingFrame, bool reverse)
            {
                _stackTrace = stackTrace;
                _startingFrame = startingFrame;
                _endingFrame = endingFrame - 1;
                _reverse = reverse;
            }

            public IEnumerator<StackFrame> GetEnumerator()
            {
                for (int i = 0; i < Count; ++i)
                    yield return this[i];
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return GetEnumerator();
            }
        }
```

File: src/NLog/SetupExtensionsBuilderExtensions.cs (hunk 2 above)
Comment (AlanLiu90) [reply src/NLog/SetupExtensionsBuilderExtensions.cs:148]: OK, but how do I remove this commit from current PR? Revert and commit?

File: src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs [hunk 4]
```diff
@@ -132,141 +135,122 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
...
+                builder.Append(f.ToString());
             }
         }
 
-        private void AppendDetailedFlatReversed(StringBuilder builder, LogEventInfo logEvent, int startingFrame, int endingFrame)
+        private void AppendFlat(StringBuilder builder, StackFrameList stackFrameList)
         {
             bool first = true;
-            for (int i = startingFrame; i <= endingFrame; ++i)
+            foreach (var f in stackFrameList)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs:205]: To reduce memory allocations, then please do like this:

```c#
for (int i = 0; i < stackFrameList.Count; ++i)
{
      var f = stackFrameList[i];
```

File: src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs [hunk 5]
```diff
@@ -132,141 +135,122 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
...
         }
 
-        private bool AppendDetailedFlat(StringBuilder builder, LogEventInfo logEvent, int i, bool first)
+        private void AppendDetailedFlat(StringBuilder builder, StackFrameList stackFrameList)
         {
-            StackFrame f = logEvent.StackTrace.GetFrame(i);
-            var method = f.GetMethod();
-            if (method == null)
+            bool first = true;
+            foreach (var f in stackFrameList)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs:237]: To reduce memory allocations, then please do like this:

```c#
for (int i = 0; i < stackFrameList.Count; ++i)
{
      var f = stackFrameList[i];
```

File: src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs [hunk 6]
```diff
@@ -132,141 +135,122 @@ protected override void Append(StringBuilder builder, LogEventInfo logEvent)
...
-
-            return first;
         }
 
-        private void AppendDetailedFlat(StringBuilder builder, LogEventInfo logEvent, int startingFrame, int endingFrame)
+        private static void AppendRaw(StringBuilder builder, StackFrameList stackFrameList)
         {
-            bool first = true;
-            for (int i = startingFrame; i >= endingFrame; --i)
+            foreach (var f in stackFrameList)
```
Comment (snakefoot) [inline src/NLog/LayoutRenderers/CallSite/StackTraceLayoutRenderer.cs:196]: To reduce memory allocations, then please do like this:

```c#
for (int i = 0; i < stackFrameList.Count; ++i)
{
      var f = stackFrameList[i];
```

File: src/NLog/SetupExtensionsBuilderExtensions.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/SetupExtensionsBuilderExtensions.cs:148]: I'm not very skilled with git. When I need to fix previous commits, then I go into branch history and choose "Reset" and "Keep Changes" (all reverted commits are now in pending changes). Then I make the necessary code-changes and commit again. And then finally perform a `git push -f` (forced). Ofcourse force-push is a no-go when working on shared branches, but it is fine for your own local branches used for pull-requests.

**Comment by snakefoot** [comment]
Updated wiki: https://github.com/NLog/NLog/wiki/Stack-Trace-Layout-Renderer
